### ISSUE TITLE
Akka Http Server needs to handle bad paths with status 400 and not 500

### ISSUE THREAD

**Issue by zvuki** [playframework/playframework#7359]
If I create my custom `ErrorHandler extends DefaultHttpErrorHandler` , as described here https://www.playframework.com/documentation/2.6.x/ScalaErrorHandling#extending-the-default-error-handler
I have no way to change the response payload for errors like 
```
play ! @743ica805 - Internal server error, for (POST) [/action?a=X2-10GB-LRM=] ->
play
play play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 16): a=X2-10GB-LRM=
play                ^]]
play 	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:257)
play 	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:182)
play 	at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:285)
play 	at play.core.server.Server$class.logExceptionAndGetResult$1(Server.scala:49)
play 	at play.core.server.Server$class.getHandlerFor(Server.scala:79)
play Caused by: akka.http.scaladsl.model.IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 16): sku=X2-10GB-LRM=
play                ^
play 	at akka.http.scaladsl.model.IllegalUriException$.apply(ErrorInfo.scala:40)
play 	at akka.http.scaladsl.model.Uri$.fail(Uri.scala:752)
play 	at akka.http.impl.model.parser.UriParser.fail(UriParser.scala:80)
play 	at akka.http.impl.model.parser.UriParser.parseQuery(UriParser.scala:69)
play 	at akka.http.scaladsl.model.Uri.query(Uri.scala:38)
```
which by the way should not be a 500, but 400 - it's a user mistyping a URL.

If I am getting it right, the reason is that Server.scala is statically coupled with DefaultHttpErrorHandler here:
https://github.com/playframework/playframework/blob/master/framework/src/play-server/src/main/scala/play/core/server/Server.scala#L53

**Comment by marcospereira** [comment]
Hi @zvuki, the actual application error handler is being used here:

https://github.com/playframework/playframework/blob/master/framework/src/play-server/src/main/scala/play/core/server/Server.scala#L69-L71

The default error handler is used only when, for some reason, it is not possible to the `ApplicationProvider` to get an application. So, it is not that the `Server` is tied with the `DefaultHttpErrorHandler` for all the scenarios, but only when it is not possible to start the application (and Play is still able to show a page with a nice error message).

What is happening here is that our implementation of `AkkaHttpServer` is not correctly handling the Bad Request with a 400 and, because of that, the error is being treated as a Server Error. The test for this scenario is being skipped, by the way:

https://github.com/playframework/playframework/blob/7970d9113390c7bf071ff071ac8f47fdbd9f5104/framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala#L58-L69

Long story short, this is something we need to fix in `AkkaHttpServer`.

**Comment by marcospereira** [comment]
A complete stack trace for the error:

```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 14): a=X2-10GB-LRM=
             ^]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:257)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:182)
	at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:285)
	at play.core.server.Server$class.logExceptionAndGetResult$1(Server.scala:53)
	at play.core.server.Server$class.getHandlerFor(Server.scala:83)
	at play.core.server.AkkaHttpServer.getHandlerFor(AkkaHttpServer.scala:42)
	at play.core.server.AkkaHttpServer.getHandler(AkkaHttpServer.scala:191)
	at play.core.server.AkkaHttpServer.play$core$server$AkkaHttpServer$$handleRequest(AkkaHttpServer.scala:171)
	at play.core.server.AkkaHttpServer$$anonfun$6.apply(AkkaHttpServer.scala:93)
	at play.core.server.AkkaHttpServer$$anonfun$6.apply(AkkaHttpServer.scala:93)
Caused by: akka.http.scaladsl.model.IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 14): a=X2-10GB-LRM=
             ^
	at akka.http.scaladsl.model.IllegalUriException$.apply(ErrorInfo.scala:40)
	at akka.http.scaladsl.model.Uri$.fail(Uri.scala:751)
	at akka.http.impl.model.parser.UriParser.fail(UriParser.scala:80)
	at akka.http.impl.model.parser.UriParser.parseQuery(UriParser.scala:69)
	at akka.http.scaladsl.model.Uri.query(Uri.scala:37)
	at play.core.server.akkahttp.AkkaModelConversion$$anon$2.queryMap$lzycompute(AkkaModelConversion.scala:104)
	at play.core.server.akkahttp.AkkaModelConversion$$anon$2.queryMap(AkkaModelConversion.scala:104)
	at play.api.mvc.RequestHeader$class.queryString(RequestHeader.scala:95)
	at play.api.mvc.RequestHeaderImpl.queryString(RequestHeader.scala:367)
	at play.core.routing.Route$$anon$1$$anonfun$unapply$1.apply(GeneratedRouter.scala:30)
```

**Comment by gmethvin** [comment]
I believe this bug depends on akka/akka-http#210.

Now, the server does actually send a 400 error, but it's coming directly from akka-http. We would need to hook into the error handling somehow to be able to redirect it to Play's HttpErrorHandler.

**Comment by jrudolph** [comment]
In this case it seems the error isn't handled by akka-http at all but just thrown from there so it seems Play's `AkkaModelConversion` should be able to do whatever it wants.

For this particular problem (two `'='` characters in a query parameter) we are also currently considering to loosen the rules: https://github.com/akka/akka-http/issues/1120).

**Comment by zvuki** [comment]
To be honest I fine with the rules - malformed URLs should be declined, I am looking for returning 400 in this situation.

**Comment by gmethvin** [comment]
@jrudolph In this case you're right. I actually was testing with this test: https://github.com/playframework/playframework/blob/7970d9113390c7bf071ff071ac8f47fdbd9f5104/framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala#L58-L69. Actually, that test does return a 400 directly. But an invalid query, e.g. `?a=X2-10GB-LRM=` throws an exception, resulting in a 500 error.

If we tried to fix it in Play that would mean some types of bad requests are handled by Akka HTTP and some are not. Is there a good justification to handle this specific type of parsing error in Play rather than Akka HTTP?

**Comment by gmethvin** [comment]
I think I answered my own question already: it doesn't make sense for either Play or Akka HTTP to return a 400 error here, because the query string is technically valid per RFC3986. See my comments on akka/akka-http#1120.
